### PR TITLE
update squad metric fix dataset bug and add doc

### PR DESCRIPTION
**Pull request by smallv0221**
### PR types

 Function optimization 
### PR changes

APIs 
### Describe

update squad metric

### CODE REVIEW

File: paddlenlp/metrics/squad.py [hunk 1]
```diff
@@ -76,7 +74,7 @@ def compute_prediction(examples,
     # Let's loop over all the examples!
     for example_index, example in enumerate(examples):
         # Those are the indices of the features associated to the current example.
-        feature_indices = features_per_example[example_index]
+        feature_indices = features_per_example[example['id']]
```
Comment (LiuChiachi) [inline paddlenlp/metrics/squad.py:77]: 那`example_index`是不是就不需要了，可以把`enumerate`去掉

File: docs/data_prepare/data_preprocess.rst [hunk 2]
```diff
@@ -0,0 +1,130 @@
+================
+数据处理
```
Comment (chenxiaozeng) [inline docs/data_prepare/data_preprocess.rst:2]: 建议数据处理单拎出去作为以及目录，与数据集平行

数据处理包括3部分内容：（1）预训练模型的tokenizer，子目录写各类模型的tokenizer （2）非预训练模型，对应的数据处理有哪些 （3）数据处理pipeline，会介绍到data API，dataloder等。

tokenizer部分可参考：
https://huggingface.co/transformers/tokenizer_summary.html
https://huggingface.co/transformers/main_classes/tokenizer.html
better to list current support in the library for each of the models:  https://huggingface.co/transformers/
https://huggingface.co/transformers/internal/tokenization_utils.html

File: docs/data_prepare/data_preprocess.rst [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+================
+
+为方便用户进行数据处理。PaddleNLP的各个模块都封装了支持NLP任务的数据处理功能。
+
+基于预训练模型的数据处理
+------------------------
+
+在使用预训练模型做NLP任务时，需要加载对应的Tokenizer，PaddleNLP在 :class:`PreTrainedTokenizer` 中内置的 :func:`__call__` 方法可以实现基础的数据处理功能。下面以BertTokenizer举例说明：
```
Comment (chenxiaozeng) [inline docs/data_prepare/data_preprocess.rst]: 如果文档中暴露PreTrainedTokenizer的话，建议介绍得详细一些。否则就不暴露。其他知识点也一样
PreTrainedTokenizer可参考
https://huggingface.co/transformers/main_classes/tokenizer.html#pretrainedtokenizer

File: docs/data_prepare/data_preprocess.rst [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+之后使用处理好的数据集和 :func:`batchify_fn` 构建 :class:`paddle.io.DataLoader` ：
+
+.. code-block::
+
+    from paddle.io import DataLoader
+
+    train_data_loader = DataLoader(dataset=train_ds, batch_size=2, collate_fn=train_batchify_fn)
+
+到此，一个完整的数据准备流程就完成了。关于更多batchify方法，请移步collate。
```
Comment (guoshengCS) [inline docs/data_prepare/data_preprocess.rst]: 另外如果是想要介绍如何建立到DataLoader的完整数据准备流程，感觉BatchSampler相关是否也要简单的提及下呢，也可以包括我们的SamplerHelper。

File: docs/data_prepare/data_preprocess.rst [hunk 5]
```diff
@@ -0,0 +1,130 @@
+================
+数据处理
+================
+
+为方便用户进行数据处理。PaddleNLP的各个模块都封装了支持NLP任务的数据处理功能。
```
Comment (guoshengCS) [inline docs/data_prepare/data_preprocess.rst]: 感觉可以在这里再丰富下。类似于：dataset中通常为原始数据，需要经过一定的数据处理并进行采样组batch，而后通过dataloader为训练或预测使用，paddlenlp中为其中各环节提供了相应的功能支持。

File: docs/data_prepare/data_preprocess.rst [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                       #  'label': [1]}
+
+可以看到，数据集中的文本数据已经被处理成了模型可以接受的 *feature* 。:func:`map` 方法有一个重要的参数 :attr:`batched`，当设置为 :obj:`True` 时（默认为 :obj:`False` ），数据处理函数 :func:`trans_func` 的输入不再是单条数据，而是数据集的所有数据：
+
+.. code-block::
+
+    def convert_examples(examples, tokenizer):
+        querys = [example['query'] for example in examples]
+        titles = [example['title'] for example in examples]
+        tokenized_examples = tokenizer(text=querys, text_pair=titles])
```
Comment (guoshengCS) [inline docs/data_prepare/data_preprocess.rst]: 这里`tokenizer(text=querys, text_pair=titles])`是不是写错了呢

File: docs/data_prepare/data_preprocess.rst [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+                       #  'label': [1]}
+
+可以看到，在本例中两种实现的结果是相同的。但是在诸如阅读理解，对话等任务中，一条原始数据可能会产生多个 *feature* 的情况（参见 `run_squad.py <https://github.com/PaddlePaddle/PaddleNLP/blob/develop/examples/machine_reading_comprehension/SQuAD/run_squad.py>`__ ）通常需要将 :attr:`batched` 参数设置为 :obj:`True` 。 
+
+关于 :func:`map` 方法的其他参数和 :class:`paddlenlp.datasets.MapDataset` 的其他数据处理方法，请移步MapDataset。
+
+Batchify
+-----------
+
+PaddleNLP内置了多种collate function，可以协助用户简单的完成组batch的操作。
```
Comment (guoshengCS) [inline docs/data_prepare/data_preprocess.rst]: 这里是否要还要更严谨一些，组batch算BatchSampler完成的，collate function算整理batch中的数据的呢

File: docs/data_prepare/data_preprocess.rst [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+    print(tokenizer(text='天气不错')) # {'input_ids': [101, 1921, 3698, 679, 7231, 102], 'token_type_ids': [0, 0, 0, 0, 0, 0]}
+
+    # 句对转换（单条数据）
+    print(tokenizer(text='天气',text_pair='不错')) # {'input_ids': [101, 1921, 3698, 102, 679, 7231, 102], 'token_type_ids': [0, 0, 0, 0, 1, 1, 1]}
+
+    # 单句转换（多条数据）
+    print(tokenizer(text=['天气','不错'])) # [{'input_ids': [101, 1921, 3698, 102], 'token_type_ids': [0, 0, 0, 0]}, 
+                                          #  {'input_ids': [101, 679, 7231, 102], 'token_type_ids': [0, 0, 0, 0]}]
+
+关于 :func:`__call__` 方法的其他参数和功能，请移步PreTrainedTokenizer。
```
Comment (chenxiaozeng) [inline docs/data_prepare/data_preprocess.rst:28]: 请移步PreTrainedTokenizer，这里等有相关内容之后再引导。其它同理。

File: paddlenlp/datasets/experimental/dataset.py [hunk 9]
```diff
@@ -207,7 +208,8 @@ def _filter(self, data):
 
     def __iter__(self):
         num_samples = 0
-        for example in self.data():
+        self.data, data = itertools.tee(self.data)
```
Comment (guoshengCS) [inline paddlenlp/datasets/experimental/dataset.py:211]: 看了下[itertools.tee](https://docs.python.org/3/library/itertools.html#itertools.tee)应该也只是在内存暂存起来了，还要确认下这个全部保存是否是符合预期和可以接受的。

**Review by chenxiaozeng — APPROVED** [review]

File: paddlenlp/datasets/experimental/dataset.py (hunk 9 above)
Comment (guoshengCS) [reply paddlenlp/datasets/experimental/dataset.py:211]: 如果是希望像数据处理教程部分那样在自定义数据集场景下为用户提供一种MapDataset和IterDataset统一的构建方法，是否可以考虑为MapDataset和IterDataset都增加一个类似`from_read_fun`的classmethod，参数为用户自定义的read函数和`*args`，`**kwargs`

**Review by ZeyuChen — APPROVED** [review]
LGTM
